x86: add support for Mellanox Spectrum SN2000 Series

tk154 opened the pull request:
The Mellanox Spectrum SN2000 Series Switches are Managed Ethernet Switches with a maximum speed of 100Gb/s and up to 56 ports.
<br>

Tested on a Mellanox Spectrum SN2100 with the following specs:
- CPU: Intel ATOM x86 dual-core 2.4GHz
- RAM: 8GB
- Disk: 16GB SSD
- Ports: 16x QSFP28 100GbE, 1x 100M Mgmt Port, 1x RJ45 Serial Port
- USB: 1x mini 2.0
- Button: 1x (reset)
- LEDs: 6x
<br>


Installation:
- Create a bootable USB device (either by flashing this image onto it or another Linux distribution)
- Unzip the image (openwrt-x86-64-mellanox_spectrum-sn2000-ext4-combined.img.gz) and copy it onto the USB device
- Plug the USB device into the Mellanox Switch and boot from it
- Flash the image (e.g., with dd) onto the internal SSD of the switch (should be /dev/sda)

To enter the BIOS, reboot the switch and press CTRL+B while you see the BIOS information text (American Megatrends …). The default password to enter the BIOS is admin. To boot from the USB device, switch to the Boot index tab and set your USB device at the top of the boot order (the internal SSD should be currently there). Don't forget to set the SSD back at the top after you have flashed the image.
<br>


Commit cb6df97c23cb72b4c288b48eb150689809e18bdb adds needed packages to run the image on the switches.
Commit d3d006d67d95ac8c3e1fb517238e8b7f799bc1ac adds an x86_64 target profile entry inside the build system configuration interface.

thuehn commented:
tested on NVIDIA Mellanox SN2100 ... OpenWrt image works properly .. all interfaces with 1-100GBit/s up&running.

aparcar commented on target/linux/x86/base-files/etc/board.d/02_network:
```diff
@@ -31,6 +31,9 @@ cisco-mx100-hw)
 	ucidef_set_interfaces_lan_wan "mgmt eth2 eth3 eth4 eth5 eth6 eth7 eth8 eth9 eth10 eth11" "wan"
 	;;
+mellanox-technologies-ltd-msn2100)
+	ucidef_set_interface_lan "eth1 eth2 eth3 eth4 eth5 eth6 eth7 eth8 eth9 eth10 eth11 eth12 eth13 eth14 eth15 eth16"
```
Please have a look how Realtek switches are handled [here](https://github.com/openwrt/openwrt/blob/main/target/linux/realtek/base-files/etc/board.d/02_network). Switches use `lan` rather than `eth` for interface names. Maybe @blogic or @dangowrt could comment on this.

Another idea could be to port the `lan_list` function to something more generic and use the variable instead of manually defining the interfaces.

aparcar replied:
Nevermind this, keep it with eth.

skorpy2009 replied:
Mellanox goes with swp so maybe go to that?

tfhdx committed:
packages: add needed packages for Mellanox Spectrum Switches

Add needed kernel and firmware packages for Mellanox Spectrum
SN2000 Series Switches.

Signed-off-by: Til Kaiser <til.kaiser@gmx.de>

tfhdx committed:
x86: add support for Mellanox Spectrum SN2000 Switches

The Mellanox Spectrum SN2000 Series Switches are Managed Ethernet
Switches with a maximum speed of 100Gb/s and up to 56 ports.

Tested on a Mellanox Spectrum SN2100 with the following specs:
 - CPU: Intel ATOM x86 dual-core 2.4GHz
 - RAM: 8GB
 - Disk: 16GB SSD
 - Ports: 16x QSFP28 100GbE, 1x 100M Mgmt Port, 1x RJ45 Serial Port
 - USB: 1x mini 2.0
 - Button: 1x (reset)
 - LEDs: 6x

Installation:
 - Create a bootable USB device (either by flashing this image
   onto it or another Linux distribution)
 - Unzip the image
   (openwrt-x86-64-mellanox_spectrum-sn2000-ext4-combined.img.gz)
   and copy it onto the USB device
 - Plug the USB device into the Mellanox Switch and boot from it
 - Flash the image (e.g., with dd) onto the internal SSD
   of the switch (should be /dev/sda)

To enter the BIOS, reboot the switch and press CTRL+B while you see
the BIOS information text (American Megatrends …). The default password
to enter the BIOS is admin. To boot from the USB device, switch to the
Boot index tab and set your USB device at the top of the boot order
(the internal SSD should be currently there). Don't forget to set
the SSD back at the top after you have flashed the image.

Signed-off-by: Til Kaiser <til.kaiser@gmx.de>

blocktrron commented:
@tk154 can you merge Pauls requests? I think this looks good apart from those

tk154 commented:
I made an update with the following changes:

- Added tabs inside 64.mk (as requested by @aparcar)
- Separated the addition of required packages from the Mellanox target into an additional commit
- Renamed the target to "Mellanox Spectrum SN2000 Series" since the commit should work for all switches of the SN2000 Series
- Added installation steps to the target commit message instead of referring to the OpenWrt Wiki

blocktrron commented:
Merged to my staging tree - thanks!

aep commented:
is this really intended to be written to /dev/sda, overwriting ONIE?

aep commented:
apparantly you can just pull out the ssd and write to the directly. however, it doesnt work on my SN2100

```
[   11.194085] mlxsw_spectrum 0000:01:00.0: cannot register bus device
[   11.201180] mlxsw_spectrum: probe of 0000:01:00.0 failed with error -5

```

tk154 commented:
> is this really intended to be written to /dev/sda, overwriting ONIE?
> 
> i'm asking because for accessing the BIOS you need a valid support contract. "admin" no longer works, it asks you to call nvidia, which of course won't sell you support contracts for EOL devices, so this setup won't help to keep them out of landfill.

Yes, currently, that's intended. To work with ONIE, the OpenWrt image needs to be converted into an ONIE-compatible installer image or manually installed by editing the ONIE boot entries. But a sysupgrade would overwrite ONIE anyway, so I decided against it for now.
<br>

> apparantly you can just pull out the ssd and write to the directly. however, it doesnt work on my SN2100
> 
> ```
> [   11.194085] mlxsw_spectrum 0000:01:00.0: cannot register bus device
> [   11.201180] mlxsw_spectrum: probe of 0000:01:00.0 failed with error -5
> ```

It might be that the firmware version of your Switch is too new; this could have happened if you booted a Linux kernel version greater than OpenWrt's current 5.15 version before. You can check the current firmware version with _ethtool -i eth1_ and try to flash an older one. (I think 13.2008.2018 is for 5.15, but I cannot check it currently.)

<pre>ethtool -f eth1 mellanox/mlxsw_spectrum-13.2008.2018.mfa2</pre>

More info on that: https://github.com/Mellanox/mlxsw/wiki/Updating-Firmware. But keep in mind that the firmware will be reflashed if you boot a newer kernel version with mlxsw_spectrum again.

Thank you for testing the image.

aep commented:
thanks for your help!

> ethtool -f eth1 mellanox/mlxsw_spectrum-13.2008.2018.mfa2

this would require the device to be available.

i tried with mstfwmanager, but i cant figure out which thing to flash. the one you suggested results in:

```
Device #1:
----------

  Device Type:      Spectrum
  Part Number:      MSN2100-Bxxx_Ax
  Description:      Spectrum based 40GbE; 1U Open Ethernet Switch with MLNX-OS; 16 QSFP28 ports; Rangeley CPU; RoHS6
  PSID:             MT_3000114033
  PCI Device Name:  01:00.0
  Base MAC:         248a07cd3bc0
  Versions:         Current        Available     
     FW             13.1210.0122   N/A           

  Status:           No matching image found

-E- Could not find matching firmware images for 1 device(s).

Device #1: Error - no matching image found
-E- One or more errors were encountered.
```

tk154 commented:
I think for you, it was the opposite: your firmware version was too old.
At least version 13.1420.122 is needed for in-driver firmware flashing to work.
I haven't tried to flash firmware via external tools yet since there was already a flashing-capable firmware on our SN2100, but I will try that out soon.<br>
Thanks for bringing this to my attention!

feckert commented:
@blocktrron With my last rebase I noticed that a lot of new packages have been added for the x86/64 target, none of which I need! I know there is currently only one generic target. But maybe now is the time to change that and bring it in line with the other targets.

ynezz commented:
@blocktrron @tk154 @thuehn @aep I'm wondering if we really need to ship that 31MiB firmware in the default x86/64 image?

From the [following](https://github.com/Mellanox/mlxsw/wiki/Updating-Firmware):

> Upon driver initialization, the current firmware version used by the device is validated against the specific version supported by the driver. If the current version is different, the driver flashes its own firmware version.

to me, it doesn't look like its mandatory to have the mlx ASIC firmware in the system image, the user just needs to make sure, that the ASICs are flashed with the compatible versions kernel driver expects and it should work without that huge package nobody else needs.

ynezz commented:
There are 27 firmware files in that firmware `mlxsw_spectrum-firmware_20230804-1_x86_64.ipk` package, so it makes me wonder, which one is being actually used/needed for your device? Maybe we could split that huge package into several smaller ones instead, based on the ASIC version (SP1-SP4?) and/or what currently supported kernel expects to be available?

aep commented:
> @blocktrron @tk154 @thuehn @aep I'm wondering if we really need to ship that 31MiB firmware in the default x86/64 image?

I think it has to be in there, yes. 
Otherwise you need to install a different linux first to update the firmware. Openwrt traditionally "just works".

ynezz commented:
> I think it has to be in there, yes.

Looking into the source code of the driver I'm sure about the opposite :)

> Openwrt traditionally "just works".

Till this PR the image size was around 10MB. Now we've 40MB monster, just to support one $10k (read very niche) platform?

aep commented:
> Looking into the source code of the driver I'm sure about the opposite :)

specifically what i mean is that it's unlikely that the current firmware matches the openwrt kernel, so it just wont boot and users likely wont really know why immediately. openwrt in general tens to just work out of the box without understanding the hardware architecture.

> Till this PR the image size was around 10MB. Now we've 40MB monster, just to support one $10k (read very niche) platform?

that's a different discussion. 
switchdev devices arent just Nvidia anymore. Marvel and Broadcom also make chips.
This is the future of datacenter hardware.

I think openwrt on the class of device makes sense in general and should be a selectable target.
The current default here is Fedora, which is on shaky grounds.

but i would not be the one driving it given that i burnt myself on openwrt upstream before.

tapper82 commented:
I don't want this driver in my x86 bin. Thanks.

ynezz commented:
> openwrt in general tends to just work out of the box without understanding the hardware architecture.

IIUC then we just need a _single_ firmware file [mlxsw_spectrum-13.2010.1006.mfa2](https://github.com/openwrt/openwrt/pull/13924#discussion_r1391398603) (1.6MiB), don't we? So why do we need to ship 26 other files worth of 29MiB? Current kernel driver is not going to use any of those, so its just an oversight which needs to be fixed?

aep commented:
> a single firmware file

hmmm. this feels correct. but i'm no expert on mellanox.
the maintainer would need to update this file when the kernel changes, but i think thats fine and fairly standard for openwrt.

tk154 commented:
> a single firmware file

As long as the <code>mlxsw_spectrum-firmware</code> package stays inside the default x86 image, I agree. But, we would need two firmware files: one for the current kernel and one for the testing kernel version.<br>
If the package(s) will be removed from the default x86 image, I think it isn't worth it.

ynezz commented:
> As long as the mlxsw_spectrum-firmware package stays inside the default x86 image, I agree.

There is probably a little bit of misunderstanding here :) You've introduced this regression, so you're supposed to fix this, thanks!

> But, we would need two firmware files: one for the current kernel and one for the testing kernel version.

I don't think, that this is actually needed now, but once it is needed, then you would actually need to make two packages, one for kernel X/current and second one for kernel Y/testing and then select correct one by kernel version package dependency.

>  If the package(s) will be removed from the default x86 image, I think it isn't worth it.

Indeed, 2MiB size increase is quite huge, 20%, but x86 can handle it. Dunno if x86 user base :D

aparcar commented:
I commented on the mailing list that we should offer three z86 flavors, regular, extended and virtual. With that approach people ideally find what suits them

aep commented:
temp hack might be to just not include it and have the wiki instructions mention that the package needs to be post-installed?

ideally x86 would have a proper target selection like the other architectures, but i guess no one really wants to put in the work for that right now
